[WIP] Treat nonescaping closure types as owned types in SIL; use borrowing to model capture lifetimes

jckarter opened the pull request.

jckarter committed:
SIL: Model noescape partial_applys with ownership in OSSA.

Although nonescaping closures are representationally trivial pointers to their
on-stack context, it is useful to model them as borrowing their captures, which
allows for checking correct use of move-only values across the closure, and
lets us model the lifetime dependence between a closure and its captures without
an ad-hoc web of `mark_dependence` instructions.

During ownership elimination, We eliminate copy/destroy_value instructions and
end the partial_apply's lifetime with an explicit dealloc_stack as before,
for compatibility with existing IRGen and non-OSSA aware passes.

jckarter commented:
@swift-ci Please build toolchain

jckarter commented:
@swift-ci Please build toolchain

jckarter commented:
@swift-ci Please build toolchain

jckarter commented:
@swift-ci Please build toolchain

jckarter commented:
@swift-ci Please build Linux toolchain

jckarter commented:
@swift-ci Please build toolchain

jckarter commented:
https://github.com/apple/swift-integration-tests/pull/105

@swift-ci Please build toolchain

atrick commented:
The main question is: why do we want a nonescaping closure to be an owned value?

We have modeled owned values as introducing separate lifetimes. That means we can work with owned values independently from each other. For example, we can freely extend their lifetime during transformations to avoid copies.

Borrowed values depend on owned values, not vice-versa. If an owned value depends on the lifetime of a borrowed value, then that inverts the model and means we can't transform individual owned values anymore.

Also, we attach scope information to the value that introduces a lifetime and always identify the end of the scope with a set of direct lifetime-ending uses. We would care about this, for example, if we ever wanted to allow non-escaping closures lifetimes to be associated with a lexical variable.

We handle various of instructions that introduce borrow scopes this way, like `load_borrow` and `store_borrow`. `end_borrow` always ends the borrow scope. `begin_apply` works exactly the same, but happens to have a special `end_apply` for other reasons. While `begin_access` isn't a borrow scope, it follows the same form.

By analogy, if a partial apply introduces a borrow scope, we expect:
```
%pa1 = partial_apply(%x)
%pa2 = convert %pa1
apply(%pa2)
end_borrow %pa1
destroy_value %x
```
We could eventually make it an "implicit borrow" and ditch the `end_borrow`. The important thing is just that `%pa2` above is a guaranteed value. That's what tells us that it depends on another lifetime.

Otherwise, nothing stops transformations from innocently doing:
```
%pa1 = partial_apply(%x)
%pa2 = convert %pa1
destroy_value %x
apply(%pa2)
destroy_value %pa2
```

jckarter commented:
@atrick Although the handling is specific to partial applications here, it seems to me that the notion of an owned value that acts as a borrow scope would generalize to other lifetime-dependent types in the future, since a `BufferView` or a `struct Ref<T> { var x: borrowing T }` would also act as a value whose lifetime acts as a borrow scope for another value referenced by pointers/references inside the first value. Is there a better general way to model that?

> The ownership verifier will catch it because of its special handling of partial_apply. But there's no general way for code working with %pa2 to know that.

So far, it appears that OSSA passes are pretty disciplined about using BorrowingOperand and its analysis and benefits from the same special handling, but are there any other analyses I should keep an eye on? I had the OwnershipEliminator reintroduce `mark_dependence`s to try to keep post-OSSA passes happy with the old representation.

atrick commented:
> @atrick Although the handling is specific to partial applications here, it seems to me that the notion of an owned value that acts as a borrow scope would generalize to other lifetime-dependent types in the future, since a `BufferView` or a `struct Ref<T> { var x: borrowing T }` would also act as a value whose lifetime acts as a borrow scope for another value referenced by pointers/references inside the first value. Is there a better general way to model that?

I think owned on-stack closures *can* be made to work. I don't think the SIL representation makes sense, but the approach may make sense for now in the context of SILGen. I hope we can come back to this before long and fixup the representation during SILGenCleanup instead. I want to get back to the invariant that only guaranteed values can introduce a borrow scope. Having a value that both introduces a borrow scope and is itself consumed is a totally unexpected special case.
 
> > The ownership verifier will catch it because of its special handling of partial_apply. But there's no general way for code working with %pa2 to know that.
> 
> So far, it appears that OSSA passes are pretty disciplined about using BorrowingOperand and its analysis, and they also get the benefits from the same special handling, but are there any other analyses I should keep an eye on? I had the OwnershipEliminator reintroduce `mark_dependence`s to try to keep post-OSSA passes happy with the old representation.

You're right that we need some general strategy for handling lifetime-dependent values that can't simply be borrowed. That will be similar to what you've done here to work around the lack of a borrow scope. I imagine doing that without violating ownership. We can have lifetime-dependent owned values, which are consumed, or lifetime-dependent guaranteed values, which cannot be consumed. Everyone who touches an owned value in a way that might extend its lifetime will need to first prove that it's not lifetime-dependent. I believe lifetime-dependent values will ultimately be within a borrow scope though. That provides a way of verifying that we've always found the correct outer scope!

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test Linux

atrick reviewed (commented):
I left a series of comments on this PR a few days back that somehow didn't get submitted. They're just suggestions. No cause for concern.

jckarter committed:
AllocBoxToStack: Don't try to destroy promoted stack closure captures.

If the partial_apply is already [on_stack], then ClosureLifetimeFixup would
have already turned the captures into borrows and established their final
lifetimes.

jckarter commented on include/swift/SIL/OwnershipUtils.h:
```diff
@@ -311,6 +313,15 @@ class BorrowingOperandKind {
+    case SILInstructionKind::BuiltinInst: {
+      auto bi = cast<BuiltinInst>(i);
+      if (bi->getBuiltinKind() == BuiltinValueKind::StartAsyncLetWithLocalBuffer) {
+        return Kind::BeginAsyncLet;
```
Hm, that's also a problem for the `stackAlloc` and `stackDealloc` builtins used by `withUnsafeTemporaryAllocation` then.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter committed:
CanonicalizeOSSALifetime: Distinguish destroys of the current def

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

shahmishal commented:
@swift-ci Please test source compatibility

jckarter commented on lib/SIL/IR/SILInstruction.cpp:
```diff
@@ -1657,6 +1664,46 @@ bool SILInstruction::maySuspend() const {
+    // partial_apply that don't forward it along, aside from destroy_value.
+    assert(use->getUser()->hasResults());
+    for (auto result : use->getUser()->getResults()) {
+      if (!visitRecursivelyLifetimeEndingUses(result, noUsers, func)) {
```
I think I had that assert in at one point and it didn't cause any problems. I must've lost it in the middle of other changes. I'll try adding it back.
